src/tpm2-tss-engine-tcti.c

```diff
@@ -0,0 +1,426 @@
+
+#define TMP2TSS_TCTI_NAMEFORMAT "libtss2-tcti-%s.so"
+#define TPM2TSS_TCTI_ENVVAR "TPM2TSSENGINE_TCTI"
+
+/*  Define an explicit dl_handle type to reduce confusion */
+typedef void* dl_handle_t;
+
+static char*
+tcti_expand_dlname (const char *shortname)
+{
+    char *expanddlname;
+    /*  determine required buffer length and allocate it*/
+    int size = snprintf (NULL,
+                         0,
+                         TMP2TSS_TCTI_NAMEFORMAT,
+                         shortname);
+    if (size <= 0) {
+        ERR(tcti_expand_dlname, TPM2TSS_R_GENERAL_FAILURE);
+        expanddlname = NULL;
+    } else {
+        expanddlname = (char*)malloc((size_t)(size+1));
+        if (!expanddlname) {
+            ERR(tcti_expand_dlname, ERR_R_MALLOC_FAILURE);
+        } else {
+            int print_size = snprintf (expanddlname,
+                                       (size+1),
+                                       TMP2TSS_TCTI_NAMEFORMAT,
+                                       shortname);
+            if (print_size != size) {
+                ERR(tcti_expand_dlname, TPM2TSS_R_GENERAL_FAILURE);
+                free(expanddlname);
+                expanddlname = NULL;
+            }
+        }
+    }
+    return expanddlname;
+}
+
+static TSS2_RC
+tcti_dlopen (const char     *dl_path,
+             dl_handle_t    *dl_handle_p)
+{
+    TSS2_RC r;
+    dl_handle_t dl_handle = dlopen(dl_path, RTLD_LAZY);
+    if (dl_handle) {
+        *dl_handle_p = dl_handle;
+        r = TSS2_RC_SUCCESS;
+    } else {
+        char *expanddlname = tcti_expand_dlname(dl_path);
+        if (!expanddlname) {
+            ERR(tcti_dlopen, TPM2TSS_R_GENERAL_FAILURE);
+            r = TSS2_BASE_RC_GENERAL_FAILURE;
+        } else {
+            dl_handle = dlopen(expanddlname, RTLD_LAZY);
+            if (dl_handle) {
+                *dl_handle_p = dl_handle;
+                r = TSS2_RC_SUCCESS;
+            } else {
+                ERR(tcti_dlopen, TPM2TSS_R_DL_OPEN_FAILED);
+                r = TSS2_BASE_RC_BAD_REFERENCE;
+            }
+            free(expanddlname);
+        }
+    }
+    return r;
+}
+
+/*  Given the handle of a loaded TCTI library, get the pointer to the
+    TCTI-initialization function. */
+static TSS2_RC
+tcti_get_init (dl_handle_t          dl_handle,
+               TSS2_TCTI_INIT_FUNC  *init_p)
+{
+    TSS2_RC r;
+    TSS2_TCTI_INFO_FUNC getinfo =
+        (TSS2_TCTI_INFO_FUNC)dlsym(dl_handle, TSS2_TCTI_INFO_SYMBOL);
+    if (!getinfo) {
+        ERR(tcti_get_init, TPM2TSS_R_DL_INVALID);
+        r = TSS2_BASE_RC_BAD_REFERENCE;
+    } else {
+        const TSS2_TCTI_INFO *info_p = getinfo();
+        *init_p = info_p->init;
+        r = TSS2_RC_SUCCESS;
+    }
+    return r;
+}
+
+static void
+tcti_dlclose (dl_handle_t *dl_handle_p)
+{
+    if (dl_handle_p && *dl_handle_p) {
+#ifndef DISABLE_DLCLOSE
+        dlclose(*dl_handle_p);
+#endif
+        *dl_handle_p = NULL;
+    }
+}
+
+
+/*  Alignment macros used by tcti_buf_alloc */
```

Comment:
Thanks for reviewing the code, @Andreas.

My goal was to keep the overall behavior similar to the current behavior of `Esys_Initialize` and `Esys_Finalize` (but inside out). My preference would be to keep it clean and symmetric, and not leak the `dlhandle`.

I understand your concerns about the alignment. Just to confirm, are you proposing using a wrapper structure (lets say `ectxw`) that stores the DL handle and esys context separately? Then, at every Esys_... API call, access that structure member?

For example, replace:
https://github.com/tpm2-software/tpm2-tss-engine/blob/c9061a7d842e5bbe669fc2f84b9c58e8e47b1743/src/tpm2-tss-engine-rsa.c#L160
` ... `
https://github.com/tpm2-software/tpm2-tss-engine/blob/c9061a7d842e5bbe669fc2f84b9c58e8e47b1743/src/tpm2-tss-engine-rsa.c#L164-L167

with

```
    r = init_tpm_key(&**ectxw**, &keyHandle, tpm2Data);
    ...
    r = Esys_RSA_Decrypt(ectxw->etcx, keyHandle,
                         ESYS_TR_PASSWORD, ESYS_TR_NONE, ESYS_TR_NONE,
                         &digest, &inScheme, &label,
                         &sig);
```

That sounds reasonable.

Reply:
I guess, I do... ;-)
Though this also looks a bit ugly, but yes...
Maybe one change would be to stack-allocate the ectxw instead of heap-allocating.
Don't know right now, if this works with the actual code though.